src/main/java/net/neoforged/bus/ListenerList.java

```diff
@@ -30,247 +29,109 @@
+    private List<ListenerList> children;
+    private Semaphore writeLock = new Semaphore(1, true);
 
-    public ListenerList()
-    {
-        this(null);
-    }
+    ListenerList() {
+        int count = EventPriority.values().length;
+        priorities = new ArrayList<>(count);
 
-    public ListenerList(@Nullable ListenerList parent)
-    {
-        // parent needs to be set before resize !
-        this.parent = parent;
-        extendMasterList(this);
-        resizeLists(maxSize);
-    }
-
-    private synchronized static void extendMasterList(ListenerList inst)
-    {
-        allLists.add(inst);
-    }
-
-    static void resize(int max)
-    {
-        if (max > maxSize)
-        {
-            synchronized (ListenerList.class)
-            {
-                if (max > maxSize)
-                {
-                    allLists.forEach(list -> list.resizeLists(max));
-                    maxSize = max;
-                }
-            }
+        for (int x = 0; x < count; x++) {
+            priorities.add(new ArrayList<>());
         }
     }
 
-    private synchronized void resizeLists(int max)
-    {
-        if (parent != null)
-        {
-            parent.resizeLists(max);
-        }
-
-        if (lists.length >= max)
-        {
-            return;
-        }
 
-        ListenerListInst[] newList = new ListenerListInst[max];
-        int x = 0;
-        for (; x < lists.length; x++)
-        {
-            newList[x] = lists[x];
-        }
-        for(; x < max; x++)
-        {
-            if (parent != null)
-            {
-                newList[x] = new ListenerListInst(parent.getInstance(x));
-            }
-            else
-            {
-                newList[x] = new ListenerListInst();
-            }
-        }
-        lists = newList;
+    ListenerList(ListenerList parent) {
+        this();
+        this.parent = parent;
+        this.parent.addChild(this);
     }
 
-    public static synchronized void clearBusID(int id)
-    {
-        for (ListenerList list : allLists)
-        {
-            list.lists[id].dispose();
-        }
+    /**
+     * Returns a ArrayList containing all listeners for this event,
+     * and all parent events for the specified priority.
+     * <p>
+     * The list is returned with the listeners for the children events first.
+     *
+     * @param priority The Priority to get
+     * @return ArrayList containing listeners
+     */
+    public ArrayList<IEventListener> getListeners(EventPriority priority) {
+        writeLock.acquireUninterruptibly();
+        ArrayList<IEventListener> ret = new ArrayList<>(priorities.get(priority.ordinal()));
+        writeLock.release();
+        if (parent != null) {
+            ret.addAll(parent.getListeners(priority));
+        }
+        return ret;
     }
 
-    protected ListenerListInst getInstance(int id)
-    {
-        return lists[id];
+    /**
+     * Returns a full list of all listeners for all priority levels.
```

Comment:
```suggestion
     * {@return a full list of all listeners for all priority levels}
```
and remove the @return declaration.

Reply:
I can change this if you want, but I just moved the code that existed before to a different indentation level.